src/runtime_src/hip/api/hip_context.cpp

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023-2024 Advanced Micro Device, Inc. All rights reserved.
+
+#include "hip/config.h"
+#include "hip/hip_runtime_api.h"
+
+#include "hip/core/common.h"
+#include "hip/core/context.h"
+#include "hip/core/device.h"
+
+namespace xrt::core::hip {
+// Returns handle to context
+// Throws on error
+static context_handle
+hip_ctx_create(unsigned int flags, hipDevice_t device)
+{
+  auto hip_dev = device_cache.get(static_cast<device_handle>(device));
+  if (!hip_dev)
+    throw xrt_core::system_error(hipErrorInvalidValue, "device requested is not available");
+
+  hip_dev->set_flags(flags);
+  auto hip_ctx = std::make_shared<context>(hip_dev);
+  tls_objs.ctx_stack.push(hip_ctx);  // make it current
+  context_cache.add(hip_ctx.get(), std::move(hip_ctx));
+  return hip_ctx.get();
+}
+
+static void
+hip_ctx_destroy(hipCtx_t ctx)
+{
+  auto handle = reinterpret_cast<context_handle>(ctx);
+  if (!handle) {
+    throw xrt_core::system_error(hipErrorInvalidValue, "device requested is not available");
+  }
+
+  auto hip_ctx = context_cache.get(handle);
+  if (!hip_ctx)
+    throw xrt_core::system_error(hipErrorInvalidValue, "context handle not found");
+
+  // Need to remove the ctx of calling thread if its the top one
```

Comment:
```suggestion
  // Need to remove the ctx of calling thread if it is the top one
```
Could you explain the logic of the stack and the cache somewhere?
So it is possible to destroy the current context. But what happens then? And if there is no more context?
Is there a `shared_ptr` still alive for it?

Reply:
In HIP, application doesn't explicitly specify context in api calls so stack is used for getting active context. Application can call hipCtxCreate multiple times and when there is any api which needs context we check active context from stack(top one) and use it. This follows hip and cuda documentation.

Cache is map for getting actual object from handles (HIP operates on handles). We followed XRT native c++/c approach here. 
Even though application destroys Ctx using hipCtxDestroy that removed entry from map, ctx can be still alive because some other object is holding its reference and will be destroyed gracefully afterwards. This way we avoid application crashes and undefined behavior. As per cuda documentation it is undefined behavior in their case which we are avoiding in our implementation

Reply:
sure @keryell, I will add it. Thanks